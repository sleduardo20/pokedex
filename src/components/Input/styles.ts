import styled, { css } from 'styled-components';
import media from 'styled-media-query';

type ContainerProps = {
  isFocused: boolean;
  isFilled: boolean;
};

export const Container = styled.div<ContainerProps>`
  ${({ theme, isFilled }) => css`
    width: ${isFilled ? '34.8rem' : '100%'};

    ${media.lessThan('medium')`
      width: ${isFilled ? '28rem' : '100%'};
    `}

    position: relative;
    padding: 0.2rem;
    height: 4.8rem;
    background: ${theme.colors.white};
    position: relative;
    border-radius: ${theme.spacings.small};
    border: 1px solid ${isFilled ? theme.colors.ternary : theme.colors.gray};
    display: flex;
    align-items: center;
    justify-content: center;
    transition: 0.4s ease-out;

    svg {
      width: 2.4rem;
      height: 2.4rem;
      color: ${isFilled ? theme.colors.ternary : theme.colors.gray};
    }

    &:focus-within {
      width: 34.8rem;

      ${media.lessThan('medium')`
      width: 28rem;
    `}
    }

    &:focus-within label {
      font-size: 1rem;
      top: 0;
      left: 4px;
      transform: translateX(0);
    }

    button {
      width: auto;
      font-weight: ${theme.font.bold};
      cursor: pointer;
      padding: ${theme.spacings.small};
      outline: none;
      background: none;
      border: none;

      svg {
        width: 2.4rem;
        height: 2.4rem;
        color: ${isFilled ? theme.colors.ternary : theme.colors.gray};
      }
    }
    input {
      width: 100%;
      height: 95%;
      margin-top: ${theme.spacings.medium};
      outline: 0;
      border: none;
      background: none;
      padding: ${theme.spacings.small};
      font-size: ${theme.font.size.large};
      line-height: ${theme.font.size.large};
    }
    label {
      width: auto;
      position: absolute;
      top: ${isFilled ? '0' : '10px'};
      left: ${isFilled ? '4px' : '50%'};
      transform: ${isFilled ? 'translateX(0)' : 'translateX(-50%)'};
      font-size: ${isFilled ? '1rem' : '1.4rem'};
      color: ${isFilled ? theme.colors.ternary : theme.colors.gray};
      transition: all 0.2s ease-in-out;
      pointer-events: none;
    }
  `}
`;
