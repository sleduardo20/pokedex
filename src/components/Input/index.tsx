import {
  InputHTMLAttributes,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';

import { useField } from '@unform/core';

import * as S from './styles';

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  name: string;
}

const Input: React.FC<InputProps> = ({ name, placeholder, ...rest }) => {
  const inputRef = useRef<HTMLInputElement>(null);

  const [isFocused, setIsFocused] = useState(false);
  const [isFilled, setIsFilled] = useState(false);
  const [passwordShow, setPasswordShow] = useState(false);

  const { fieldName, registerField } = useField(name);

  useEffect(() => {
    registerField({
      name: fieldName,
      ref: inputRef.current,
      path: 'value',
    });
  }, [registerField, fieldName]);

  const handelOnFocus = useCallback(() => {
    setIsFocused(true);
  }, []);

  const handleBlur = useCallback(() => {
    setIsFocused(false);
  }, []);

  const handleFilled = useCallback(text => {
    if (text) {
      setIsFilled(true);
    } else {
      setIsFilled(false);
    }
  }, []);

  const handleShowPassword = useCallback(() => {
    setPasswordShow(state => !state);
  }, []);

  return (
    <S.Container isFocused={isFocused} isFilled={isFilled}>
      <label htmlFor={name}>{placeholder}</label>
      <input
        ref={inputRef}
        name={name}
        onFocus={handelOnFocus}
        onBlur={handleBlur}
        type={passwordShow ? 'text' : 'password'}
        onChange={e => handleFilled(e.target.value)}
        {...rest}
      />

      {name === 'password' && isFilled && (
        <button onClick={handleShowPassword} type="button">
          {passwordShow ? 'Ocultar' : 'Mostrar'}
        </button>
      )}
    </S.Container>
  );
};

export default Input;
