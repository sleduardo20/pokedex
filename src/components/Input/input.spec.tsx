import { screen } from '@testing-library/react';

import { renderWithTheme } from 'util/renderWithTheme';
import { Form } from '@unform/web';

import Input from '.';

describe('Input', () => {
  it('should be able render Input correctly', () => {
    renderWithTheme(
      <Form
        onSubmit={() => {
          console.log('ok');
        }}
      >
        <Input name="email" type="text" />
      </Form>,
    );

    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });
});
