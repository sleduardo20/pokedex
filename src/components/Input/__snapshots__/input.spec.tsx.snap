// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should be able render Input correctly 1`] = `
.c0 {
  width: 100%;
  position: relative;
  padding: 0.2rem;
  height: 4.8rem;
  background: #FFFFFF;
  position: relative;
  border-radius: 0.4rem;
  border: 1px solid #DFDFDF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: 0.4s ease-out;
  transition: 0.4s ease-out;
}

.c0:focus-within {
  width: 34.8rem;
}

.c0:focus-within label {
  font-size: 1rem;
  top: 0;
  left: 4px;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
}

.c0 button {
  width: auto;
  font-weight: 800;
  cursor: pointer;
  padding: 0.4rem;
  outline: none;
  background: none;
  border: none;
}

.c0 button svg {
  width: 2.4rem;
  height: 2.4rem;
  color: #DFDFDF;
}

.c0 input {
  width: 100%;
  height: 95%;
  outline: 0;
  border: none;
  background: none;
  padding: 0.4rem;
  font-size: 1.2rem;
  line-height: 1.6rem;
}

.c0 label {
  width: auto;
  position: absolute;
  top: 10px;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  font-size: 1.4rem;
  color: #DFDFDF;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  pointer-events: none;
}

<form>
  <div
    class="c0"
  >
    <label
      for="email"
    >
      Usuario ou Email
    </label>
    <input
      name="email"
      type="text"
    />
  </div>
</form>
`;
